### PR TITLE
Add the consistency check tool

### PR DESCRIPTION
**Pull request by SimFG**
#16997 
Signed-off-by: SimFG <bang.fu@zilliz.com>

### CODE REVIEW

File: cmd/milvus/help.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+`
+	stopLine = `
+milvus stop [server type] [flags]
+	Stop a Milvus Server.
+[flags]
+	-alias ''
+		Set alias
+`
+	mckLine = `
+milvus mck [flags]
```
Comment (jaime0815) [inline cmd/milvus/help.go:43]: suggest adding a second parameter, for example : milvus mck run ?

File: cmd/milvus/mck.go [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+	flagStartIndex int
+}
+
+func (c *mck) execute(args []string, flags *flag.FlagSet) {
+	c.initParam()
+	flags.Usage = func() {
+		fmt.Fprintln(os.Stderr, mckLine)
+	}
+	c.flagStartIndex = 2
+	isClean := len(args) >= 3 && args[2] == "clean"
```
Comment (jaime0815) [inline cmd/milvus/mck.go:77]: we'd better unify the parameter format

File: cmd/milvus/help.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	-minioUsername ''
+		The username to login the minio server.
+	-minioPassword ''
+		The password to login the minio server.
+	-minioUseSSL 'false'
+		Whether to use the ssl to connect the minio server.
+	-minioBucketName ''
+		The bucket to operate the data in it
+
+milvus mck clean [flags]
```
Comment (jaime0815) [inline cmd/milvus/help.go:62]: Use cleanTrash

File: cmd/milvus/mck.go [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+
+	// log all tasks
+	if len(c.taskNameMap) > 0 {
+		log.Info("all tasks")
+		for taskID, taskName := range c.taskNameMap {
+			log.Info("task info", zap.String("name", taskName), zap.Int64("id", taskID))
+		}
+	}
+
+	// collect invalid tasks related with invalid partitions
```
Comment (jaime0815) [inline cmd/milvus/mck.go:158]: Split the following code into functions

File: cmd/milvus/mck.go [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+		storage.BucketName(getNotEmptyString(c.minioBucketName, c.params.MinioCfg.BucketName)),
+		storage.CreateBucket(true))
+
+	var err error
+	c.minioChunkManager, err = chunkManagerFactory.NewVectorStorageChunkManager(context.Background())
+	if err != nil {
+		log.Fatal("failed to connect to etcd", zap.Error(err))
+	}
+
+	/*option := &miniokv.Option{
```
Comment (jaime0815) [inline cmd/milvus/mck.go:295]: remove these code

File: cmd/milvus/mck.go [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+	c.connectEctd()
+	if isClean {
+		c.cleanTrash()
+		return
+	}
+	c.connectMinio()
+
+	_, values, err := c.etcdKV.LoadWithPrefix(segmentPrefix)
+	if err != nil {
+		log.Fatal("failed to list the segment info", zap.Error(err))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:99]: error information should contain key

File: cmd/milvus/mck.go [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+	}
+	for _, value := range values {
+		info := &datapb.SegmentInfo{}
+		err = proto.Unmarshal([]byte(value), info)
+		if err != nil {
+			continue
```
Comment (jaime0815) [inline cmd/milvus/mck.go:105]: add a warn log

File: cmd/milvus/mck.go [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+		}
+		c.segmentIDs[info.ID] = info
+	}
+
+	_, values, err = c.etcdKV.LoadWithPrefix(collectionPrefix)
+	if err != nil {
+		log.Fatal("failed to list the collection info", zap.Error(err))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:112]: error information should contain key

File: cmd/milvus/mck.go [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+	}
+	for _, value := range values {
+		collInfo := pb.CollectionInfo{}
+		err = proto.Unmarshal([]byte(value), &collInfo)
+		if err != nil {
+			continue
```
Comment (jaime0815) [inline cmd/milvus/mck.go:118]: add a warn log

File: cmd/milvus/mck.go [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+			c.partitionIDs[id] = nil
+		}
+	}
+
+	// log Segment ids and partition ids
+	var ids []int64
+	for id := range c.segmentIDs {
+		ids = append(ids, id)
+	}
+	log.Info("segment ids", zap.Int64s("ids", ids))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:130]: log.Info("segment ids", zap.Int64s("ids", c.segmentIDs)) still can replace 126-130 ?

File: cmd/milvus/mck.go [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+
+	ids = []int64{}
+	for id := range c.partitionIDs {
+		ids = append(ids, id)
+	}
+	log.Info("partition ids", zap.Int64s("ids", ids))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:136]: the same with the abve

File: cmd/milvus/mck.go [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+
+	rootPath := getNotEmptyString(c.ectdRootPath, c.params.EtcdCfg.MetaRootPath)
+	c.etcdKV = etcdkv.NewEtcdKV(etcdCli, rootPath)
+	log.Info("Etcd root path", zap.String("root_path", rootPath))
+}
+
+func (c *mck) connectMinio() {
+	useSSL := c.params.MinioCfg.UseSSL
+	if c.minioUseSSL == "true" || c.minioUseSSL == "false" {
+		useSSL, _ = strconv.ParseBool(c.minioUseSSL)
```
Comment (jaime0815) [inline cmd/milvus/mck.go:278]: catch and process error

File: cmd/milvus/mck.go [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+	}
+	chunkManagerFactory := storage.NewChunkManagerFactory("local", "minio",
+		storage.RootPath(c.params.LocalStorageCfg.Path),
+		storage.Address(getNotEmptyString(c.minioAddress, c.params.MinioCfg.Address)),
```
Comment (jaime0815) [inline cmd/milvus/mck.go:282]: firstly,  optional parameters have higher priority,  we should check parameters' integrity once use them.
then try to use default configurations from milvus.yaml

File: cmd/milvus/mck.go [hunk 14]
```diff
@@ -0,0 +1,692 @@
...
+		return
+	}
+	fmt.Println(strings.Join(keys, "\n"))
+	fmt.Print("Delete all backup infos, [Y/n]:")
+	deleteAll := ""
+	fmt.Scanln(&deleteAll)
+	if deleteAll == "Y" {
+		err = c.etcdKV.RemoveWithPrefix(MckTrash)
+		if err != nil {
+			log.Error("failed to remove backup infos", zap.Error(err))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:334]: print key information within log

File: cmd/milvus/mck.go [hunk 15]
```diff
@@ -0,0 +1,692 @@
...
+func getTrashKey(taskType, key string) string {
+	return fmt.Sprintf("%s/%s/%s", MckTrash, taskType, key)
+}
+
+func (c *mck) extractTask(prefix string, keys []string, values []string) {
+
+	for i := range keys {
+		taskID, err := strconv.ParseInt(filepath.Base(keys[i]), 10, 64)
+		if err != nil {
+			log.Warn("failed to parse int", zap.String("tasks", filepath.Base(keys[i])))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:361]: print key information within log

File: cmd/milvus/mck.go [hunk 16]
```diff
@@ -0,0 +1,692 @@
...
+		c.taskNameMap[taskID] = taskName
+	}
+}
+
+func (c *mck) removeTask(invalidTask int64) bool {
+	taskType := c.taskNameMap[invalidTask]
+	key := c.taskKeyMap[invalidTask]
+	err := c.etcdKV.Save(getTrashKey(taskType, key), c.allTaskInfo[key])
+	if err != nil {
+		log.Warn("failed to backup task", zap.Int64("task_id", invalidTask), zap.Error(err))
```
Comment (jaime0815) [inline cmd/milvus/mck.go:388]: print key information within log

File: cmd/milvus/mck.go [hunk 17]
```diff
@@ -0,0 +1,692 @@
...
+					continue
+				}
+				if !c.removeTask(invalidTask) {
+					continue
+				}
+				delete(idMap, invalidTask)
+				fmt.Println("Delete task id: ", invalidTask)
+			}
+		}
+	}
```
Comment (jaime0815) [inline cmd/milvus/mck.go:691]: print log to all tasks are valid

File: cmd/milvus/milvus.go [hunk 18]
```diff
@@ -3,339 +3,59 @@ package milvus
...
-	// use auto max procs to set container CPU quota
-	"go.uber.org/zap"
-
-	"github.com/milvus-io/milvus/internal/log"
-	"github.com/milvus-io/milvus/internal/util/metricsinfo"
 )
 
 const (
-	roleMixture = "mixture"
+	DryRunCmd = "dry-run"
```
Comment (jaime0815) [inline cmd/milvus/milvus.go:10]: why add a dry run, it seems does nothing

File: cmd/milvus/mck.go (hunk 17 above)
Comment (SimFG) [reply cmd/milvus/mck.go:691]: All tasks have printed in the front code.

File: cmd/milvus/milvus.go (hunk 18 above)
Comment (SimFG) [reply cmd/milvus/milvus.go:10]: This was previously included.

**Comment by SimFG** [comment]
rerun ut

**Pull request merged**